File: src/main/java/com/example/cqrs/schedule/EventReplayScheduler.java
package com.example.cqrs.schedule;

import com.example.cqrs.service.AccountQueryService;
import com.example.cqrs.service.EventReplayService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.List;

@RequiredArgsConstructor
@Slf4j
@Component
public class EventReplayScheduler {

    private final EventReplayService eventReplayService;
    private final AccountQueryService accountQueryService;

    @Scheduled(cron = "0 0 0 * * *")
    public void syncDailyData() {
        log.info("Starting daily data synchronization...");
        LocalDateTime yesterday = LocalDateTime.now().minusDays(1);

        try {
            List<String> activeAccounts = accountQueryService.getActiveAccountIds();
            log.info("Found {} active accounts to sync", activeAccounts.size());

            activeAccounts.forEach(accountId -> {
                try {
                    eventReplayService.replayEvents(accountId, yesterday);
                    log.debug("Successfully synced account: {}", accountId);
                } catch (Exception e) {
                    log.error("Failed to sync account: {}", accountId, e);
                }
            });

            log.info("Completed daily data synchronization");
        } catch (Exception e) {
            log.error("Failed to execute daily sync", e);
        }
    }

}

File: src/main/java/com/example/cqrs/interfaces/scheduler/eventreplay/EventReplayScheduler.kt
package com.example.cqrs.interfaces.scheduler.eventreplay

import com.example.cqrs.application.account.command.service.usecase.AccountEventStoreUseCase
import com.example.cqrs.application.account.event.retry.EventReplayService
import com.example.cqrs.application.account.query.service.usecase.AccountUseCase
import com.example.cqrs.infrastructure.eventstore.entity.base.AccountEventBaseEntity
import com.example.cqrs.infrastructure.eventstore.entity.enumerate.EventStatus
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import java.time.LocalDateTime

@Component
class EventReplayScheduler(
    private val eventReplayService: EventReplayService,
    private val accountUseCase: AccountUseCase,
    private val accountEventStoreUseCase: AccountEventStoreUseCase
) {

    @Scheduled(cron = "0 0 0 * * *") // 매일 자정
    fun syncDailyEvents() {
        val yesterday = LocalDateTime.now().minusDays(1)

        try {
            val activeAccountIds = accountUseCase.getActiveAccountIds()

            activeAccountIds.forEach { accountId ->
                val findUnprocessedEvents = findUnprocessedEvents(accountId, yesterday)

                if (findUnprocessedEvents.isNotEmpty()) {
                    eventReplayService.replayEvents(accountId, yesterday)
                }
            }
        } catch (e: Exception) {
            // 예외 처리
        }
    }

    /**
     * 지정된 날짜 이후에 처리되지 않은 이벤트를 조회합니다.
     *
     * @param accountId 계좌 ID
     * @param fromDate  조회 시작 날짜
     * @return 처리되지 않은 이벤트 목록
     */
    private fun findUnprocessedEvents(
        accountId: String,
        fromDate: LocalDateTime
    ): List<AccountEventBaseEntity> {
        return accountEventStoreUseCase.getEvents(accountId, fromDate)
            .filter { event -> event.status != EventStatus.PROCESSED }
    }

}